### PR TITLE
allow @differentiable on protocol requirements

### PR DESCRIPTION
**Pull request by marcrasi**
The parser already allows `@differentiable` on protocol requirements, but then it crashes in SILGen when it tries to find the original function. This PR fixes that crash.

This PR also forbids custom primals and adjoints on protocol requirements so that we don't have to deal with them for now.

### CODE REVIEW

File: lib/SILGen/SILGen.cpp [hunk 1]
```diff
@@ -794,49 +794,54 @@ void SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
...
-      if (auto *primFn = diffAttr->getPrimalFunction())
-        primName = getFunction(SILDeclRef(primFn), ForDefinition)->getName();
-      if (auto *adjointFn = diffAttr->getAdjointFunction()) {
-        // If the adjoint is specified but the primal is not, then we treat the
-        // original as the primal.
-        if (primName.empty())
-          primName = silOriginalFn->getName();
-        adjName = getFunction(SILDeclRef(adjointFn), ForDefinition)->getName();
-        hasPrimitiveAdjoint = true;
+  if (AFD->hasBody()) {
```
Comment (harlanhaskins) [inline lib/SILGen/SILGen.cpp]: Drive-by nitpick: Can you invert this condition and early-return if there is no body?

Comment (rxwei) [reply lib/SILGen/SILGen.cpp]: +1

**Comment by rxwei** [comment]
> This PR also forbids custom primals and adjoints on protocol requirements so that we don't have to deal with them for now.

I don't think these cases should ever be allowed, so this change is fine.

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -2693,6 +2693,10 @@ ERROR(differentiable_attr_no_parameters,none,
       "%0 has no parameters to differentiate with respect to", (DeclName))
 ERROR(differentiable_attr_void_result,none,
       "cannot differentiate void function %0", (DeclName))
+ERROR(differentiable_attr_primal_no_definition,none,
+      "cannot specify primal on function declaration without definition", ())
```
Comment (rxwei) [inline include/swift/AST/DiagnosticsSema.def]: The only case of functions not having a definition in Swift is protocol requirements :) So we should call it "protocol requirements" instead of "function declaration without definition" like other diagnostics.

**Review by rxwei — APPROVED** [review]

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow
